typescript/introduction/tsc-classes-and-interfaces/tsc-classes-and-interfaces-intro.md

```diff
@@ -0,0 +1,127 @@
+---
+author: nem035
+type: normal
+category: must-know
+aspects:
+  - introduction
+links:
+  - '[Play with interfaces and classes](https://www.typescriptlang.org/play/index.html#code/JYOwLgpgTgZghgYwgAgGIBsCe1kG8BQyyA9McnOusjFtAM7IC2ArnWMgBZwBuKcyIOIwiESZMF3bAGAexjIwmAA4o2UUAHNRg4QC5kazQG5RpcpWq0oDFm048+lzEwgSZAE1PjJ9hiBnISnBQQq705CCeRGbuMhB0IADk7FCuzFAgEZgSmqI0mAAUAJT63DLA7iYAvvj4COhwdAwAQsBQ7sjAjEroEMLgDBjYUHjaofqGIBomRAgyIGrMCGAyUAU6EBNg6lNFo0REOXQAdBvIALwCoTPINUT5xfsHcwsyvcfoMhoFogfIAAYAIzaHQAJLgjqdQlVOgx8pp-r9kEUbjUanUGk1kAAREIaeadbq9fpgQZWJ4bLY7aaiF6LZardbjAzbTR7Ah-SFnS4bVF5LCPDnPeZ0N4QD5fH5-Ij-dx4gngrnQ2FOBFIlGiNG1OnsCBwDS9C4CCAAd2QrXaBQARABRfW9K0anXIABWcEBgOgJpkCAA1s4eaacfKQFLkFaAFLuz1Qb1+zBW-Aa-B6g3ih4aszA9rIO1plXwqb4N0er0+-3HDMmGIh5BR0ux8vOaSqotAA){documentation}'
+
+---
+
+# Classes and Interfaces
+
+---
+## Content
+
+Interfaces can be used with classes as their **behavioral contract**.
+
+For example, for a `Bird` and a `Dragon` to become flyers, they must implement the `Flyer` contract.
+
+```ts
+interface Flyer {
+  // all flyers must have a name
+  // that is of type string
+  name: string;
+  // all flyers must have a fly method
+  // that has no parameters and
+  // doesn't return anything
+  fly(): void;
+}
+
+class Bird implements Flyer {
+  name: string;
+  constructor(name: string) {
+    this.name = name;
+  }
+  fly() {
+    console.log(
+      `bird ${this.name} is flying`
+    );
+  }
+}
+
+class Dragon implements Flyer {
+  name: string;
+  constructor(name: string) {
+    this.name = name;
+  }
+  fly() {
+    console.log(
+      `dragon ${this.name} is flying`
+    );
+  }
+}
+
+const eagle = new Bird("Eagle");
+const jabberwocky = new Dragon(
+  "Jabberwocky"
+);
+
+eagle.fly();
+// bird Eagle is flying
+jabberwocky.fly();
+// dragon Jabberwocky is flying
+```
+
+
+---
+## Practice
+
+Fill in the gaps such that the code contains no type errors:
+
+```ts
+??? MobileApp {
+  category: string;
+}
+
+??? EnkiApp ??? MobileApp {
+  category: ???
```

Comment:
I didn't get from the insight that you have to do `category: string = 'Education'`

also, why wouldn't just `category.string` not work?

Reply:
This was on purpose.

There's even a hint in the code :)

It doesn't work because we're calling `new EnkiApp();` and never set the category.

<img width="619" alt="Screen Shot 2020-05-01 at 11 09 09 AM" src="https://user-images.githubusercontent.com/9661806/80815741-37782080-8b9c-11ea-838c-2180d132f39c.png">

Reply:
This was explained in the insight about classes.

The `type` field on the class `Animal` has a default value.

I'll elaborate on it a bit more in that insight then.